h/sentry/subscribers.py

```diff
@@ -0,0 +1,70 @@
+from __future__ import unicode_literals
+
+import traceback
+
+from pyramid.events import subscriber
+from pyramid_retry import IBeforeRetry
+import sentry_sdk
+
+
+@subscriber(IBeforeRetry)
+def add_retryable_error_to_sentry_context(event):
+    """
+    Add information about a retryable error to the Sentry context.
+
+    If a request raises a retryable error then, if we're not already on this
+    request's last retry attempt, pyramid_retry calls this IBeforeRetry
+    subscriber function before it retries the request.
+
+    This function adds some information to the Sentry context about the
+    retryable error that was raised by the failed attempt at the request.
+
+    If a future retry of this request raises a non-retryable error, or if we
+    use up the maximum number of retry attempts, so that we ultimately end up
+    sending an error response back to the client, then the information that
+    this function adds to the Sentry context will be included in the event that
+    is ultimately reported to Sentry.
+
+    This makes debugging errors easier because you can see information about
+    previous failed attempts at the request, as well as the usual information
+    about the last exception that was raised by the final attempt.
+
+    If a future retry of this request succeeds and we ultimately end up sending
+    a successful response back to the client then nothing will be reported to
+    Sentry.
+
+    See:
+
+    * https://docs.pylonsproject.org/projects/pyramid-retry/en/latest/#receiving-retry-notifications
+    * https://docs.sentry.io/platforms/python/#extra-context
+    """
+    attempt = event.environ.get("retry.attempt")
+    attempts = event.environ.get("retry.attempts")
+    exception = event.request.exception
+
+    if attempt is None or attempts is None or exception is None:
```

Comment:
The `or exception is None` here is now unnecessary if you access `exception` as `event.exception` rather than `event.request.exception`, and if we upgrade to the just-released [pyramid_tm 2.3](https://github.com/Pylons/pyramid_tm/blob/master/CHANGES.rst#23-2019-09-30)